### ISSUE TITLE
:union/:union-all should wrap args in parentheses

### ISSUE THREAD

**Issue by camsaul** [jkk/honeysql#279]
Related: #107

Using HoneySQL 0.9.5. I ran into a SQL syntax issue with Postgres.

### Broken:

```clj
(defn bad [field-id]
  (db/query {:union [{:select [[:dimension.human_readable_field_id :id]]
                      :from   [[Dimension :dimension]]
                      :where  [:= :dimension.field_id field-id]
                      :limit  1}
                     {:select    [[:values.id :id]]
                      :from      [[FieldValues :values]]
                      :where     [:= :values.field_id field-id]
                      :limit     1}]}))

(bad 1) 
;; -> Execution error (PSQLException) at org.postgresql.core.v3.QueryExecutorImpl/receiveErrorResponse (QueryExecutorImpl.java:2497).
;; -> ERROR: syntax error at or near "UNION" Position: 124
```

SQL:
```sql
SELECT "dimension"."human_readable_field_id" AS "id"
FROM "dimension" "dimension"
WHERE "dimension"."field_id" = 1
LIMIT 1
UNION
SELECT "values"."id" AS "id"
FROM "metabase_fieldvalues" "values"
WHERE "values"."field_id" = 1
LIMIT 1
```

### Fixed:

```clj
(defmethod honeysql.format/fn-handler "_parens" [_ x]
  (str "(" (honeysql.format/to-sql x) ")"))

(defn- parens [x]
  (hsql/call :_parens x))

(defn good [field-id]
  (db/query {:union [(parens
                      {:select [[:dimension.human_readable_field_id :id]]
                       :from   [[Dimension :dimension]]
                       :where  [:= :dimension.field_id field-id]
                       :limit  1})
                     (parens
                      {:select [[:values.id :id]]
                       :from   [[FieldValues :values]]
                       :where  [:= :values.field_id field-id]
                       :limit  1})]}))

(good 1) ; -> nil
```

SQL:

```
(SELECT "dimension"."human_readable_field_id" AS "id"
 FROM "dimension" "dimension"
 WHERE "dimension"."field_id" = 1
 LIMIT 1) 
UNION
(SELECT "values"."id" AS "id"
 FROM "metabase_fieldvalues" "values"
 WHERE "values"."field_id" = 1
 LIMIT 1)
```

**Comment by seancorfield** [comment]
It's because of `limit` -- if you remove the `:limit 1` in both `:select` queries, then it would be valid syntax (and you can `limit` the whole thing if `:limit n` is at the same level as `:union`).

Given the complexities here -- different SQL dialects having different rules about how parentheses around `union` clauses and their sub-clauses behave -- I think adding a way to explicitly wrap parens around a given construct is probably the right solution overall.

At this point, I'm not sure I want to add that to 1.0 -- and you have a workaround -- but I will add this as a "special syntax" function in 2.0 to make everyone's life easier in future.

**Comment by seancorfield** [comment]
In v2, there will be `{:nest ...}` syntax for statements and `[:nest ...]` syntax for expression to introduce a level of parentheses when the default generation of code doesn't have the nesting you need in any particular situation.

**Comment by camsaul** [comment]
I mentioned this on the Clojurians Slack as well, but I think it would be a nicer experience for everyone if we just wrap things in parens by default in cases where we might need them. It doesn't hurt to have extra parens, but it does hurt not to have them. I think it's a nicer and easier API to have it work without having to remember (or know) to use `:nest`. If people need to opt-in to parentheses with something like `:nest`, then it's going to lead to bugs for library users where it works most of the time, but fails in edge cases where parentheses would have been required

**Comment by seancorfield** [comment]
Well, you can't just wrap things in parentheses arbitrarily in SQL -- that was the original problem in #107 because there was an extra set of parens. And you don't need the parens in the vast majority of cases with union. I believe I've also run into a case, in MySQL, where if you add the parens, in some situations you also need to add an alias, even if nothing else in the query actually uses it. That sort of thing makes me very wary of adding parens by default unless they really are needed. V1 and V2 already have a different approach for this where I'm removing external parens (from where they're definitely not needed) and adding them internally where they are safe.

I'll leave this open while V2 is being worked on and see what feedback I get from other users. It's trivial to change the behavior for all the set operations without changing anything else.

**Comment by camsaul** [comment]
It sounds like the issue in #107 was caused by wrapping the entire `UNION` expression in parentheses, e.g. 

```sql
INSERT INTO baz (id, qux) ((SELECT * FROM bar) UNION (SELECT * FROM foo))
```
In #107 @tomconnors mentioned that 

```sql
INSERT INTO baz (id, qux) (SELECT * FROM bar) UNION (SELECT * FROM foo)
```
is ok. Is there any situation where wrapping only the subqueries (but not the entire `UNION` expression) would not be ok? If not, I think having this be the default behavior would make sense -- it would address this issue without causing a regression of #107

**Comment by seancorfield** [comment]
This won't get fixed in 1.0, but in 2.0 I am not treating the `UNION` sub-expressions as nested (so they get parenthesized). We'll see whether that breaks anything as people do more testing of 2.0.

**Comment by seancorfield** [comment]
I think this is correct in v2 so I'm going to close this out now.
